org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelFormatterParticipant.java

```diff
@@ -67,4 +67,23 @@ public FormatElementCategory getFormatElementCategory(DOMElement element,
 		return null;
 	}
 
+	@Override
+	public boolean shouldCollapseEmptyElement(DOMElement element, SharedSettings sharedSettings) {
+		boolean enabled = sharedSettings.getFormattingSettings().isGrammarAwareFormatting();
+		if (!enabled) {
+			return false;
+		}
+
+		Collection<CMDocument> cmDocuments = contentModelManager.findCMDocument(element);
+		for (CMDocument cmDocument : cmDocuments) {
+			CMElementDeclaration cmElement = cmDocument.findCMElement(element);
+			if (cmElement != null && (cmElement.isEmpty()
+					|| (cmElement.isNillable() && "true".equals(element.getAttribute("xsi:nil"))))) {

```

Comment:
I think this or here should be an and, right? We only want to collapse when the element doesn't have any content and the schema says that it's okay for it to be self-closing.

Reply:
I used or here because isEmpty() here checks for a specific case: https://github.com/eclipse/lemminx/blob/068e862dde1bd55539ecb8be29e1af9f854e27f5/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xsd/contentmodel/CMXSDElementDeclaration.java#L461

Perhaps it's a good idea to add the nillable conditions to this method?

Reply:
Okay. The `||` here is correct. I was confused about how `nillable` worked. I think what I ran into while testing is that I expected an element with a simple type that could be empty, such as `xs:string`, to be collapsed when formatting if it was empty.

eg. I expected `childA` to be collapsed in the following case:

```xml
<root>
  <childA></childA>
</root>
```

given the associated schema:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
  <xs:element name="root">
    <xs:complexType>
      <xs:sequence>
        <xs:element ref="childA" minOccurs="0" maxOccurs="1" />
      </xs:sequence>
    </xs:complexType>
  </xs:element>
  <xs:element name="childA">
    <xs:simpleType>
      <xs:restriction base="xs:string" />
    </xs:simpleType>
  </xs:element>
</xs:schema>
```

The issue is that `isEmpty` checks if the element MUST be empty, not if it's allowed to be empty. Unfortunately, it looks like a lot of work would be required to figure out if an elements content is allowed to be empty. So I think the implementation you have currently is good.